p2panda-rs/src/secret_group/group.rs

```diff
@@ -0,0 +1,552 @@
+
+        // Generate secret key by using the MLS exporter method
+        let value = self.mls_group.export_secret(provider, label, key_length)?;
+
+        // Store secret in internal storage
+        self.long_term_secrets.push(LongTermSecret::new(
+            self.group_instance_id().clone(),
+            self.long_term_ciphersuite,
+            long_term_epoch,
+            value.into(),
+        ));
+
+        Ok(())
+    }
+
+    // Encryption
+    // ==========
+
+    // Securely encodes and encrypts a list of long term secrets for the current MLS group. Members
+    // of this MLS group epoch will be able to decrypt and use these secrets.
+    fn encrypt_long_term_secrets(
+        &mut self,
+        provider: &impl OpenMlsCryptoProvider,
+    ) -> Result<SecretGroupMessage, SecretGroupError> {
+        // Encode all long term secrets
+        let encoded_secrets = self
+            .long_term_secrets
+            .tls_serialize_detached()
+            .map_err(|_| SecretGroupError::LTSEncodingError)?;
+
+        // Encrypt encoded secrets
+        Ok(self.encrypt(provider, &encoded_secrets)?)
+    }
+
+    // Generates unique nonce which can be used for AEAD.
+    fn generate_nonce(
+        &mut self,
+        provider: &impl OpenMlsCryptoProvider,
+    ) -> Result<Vec<u8>, SecretGroupError> {
+        let public_key_str = hex::encode(self.mls_group.credential()?.identity());
+
+        // Use constant value, public key and incrementing integer as exporter label
+        let label = &format!(
+            "{}{}{}",
+            LTS_NONCE_EXPORTER_LABEL,
+            &public_key_str,
+            self.long_term_nonce.increment(),
+        );
+
+        // Determine length of AEAD nonce.
+        let nonce_length = self.long_term_ciphersuite.aead_nonce_length();
+
+        // Retreive nonce from MLS exporter
+        let nonce = self
+            .mls_group
+            .export_secret(provider, label, nonce_length)?;
+
+        Ok(nonce)
+    }
+
+    // Decrypts and decodes a list of long term secrets received via a commit message or when
+    // joining an existing group.
+    fn decrypt_long_term_secrets(
+        &mut self,
+        provider: &impl OpenMlsCryptoProvider,
+        encrypted_long_term_secrets: SecretGroupMessage,
+    ) -> Result<LongTermSecretVec, SecretGroupError> {
+        // Decrypt long term secrets with current group state
+        let secrets_bytes = self.decrypt(provider, &encrypted_long_term_secrets)?;
+
+        // Decode secrets
+        let secrets = LongTermSecretVec::tls_deserialize(&mut secrets_bytes.as_slice())
+            .map_err(|_| SecretGroupError::LTSDecodingError)?;
+
+        Ok(secrets)
+    }
+
+    /// Encrypt user data using the a single-use "Sender Ratchet" secret generated by the current
+    /// MLS group TreeKEM algorithm.
+    ///
+    /// This method gives forward-secrecy and post-compromise security. Use this encryption method
```

Comment:
Nice suggestion, I'll take it! Just removed the `schema` word as it might be misleading in the p2panda context?

> Are there other aspects to the security guarantees offered for sender ratchet vs long-term secrets?

Think this covers the most important :+1:

> Question: can a group mix and match both methods?

Yes! Its theoretically possible.

Reply:
> Yes! Its theoretically possible. 🧠 💥

That is amazing!